src/cast_core/src/CollectionScanner.cpp

```diff
@@ -34,31 +31,127 @@
+          scanSizeBytes{context["ScanSizeBytes"].maybe<IntegerSpec>().value_or(0)} {
+        // Initialise scan type enum.
+        auto scanTypeString = context["ScanType"].to<std::string>();
+        // Ignore case
+        boost::algorithm::to_lower(scanTypeString);
+        if (scanTypeString == "count") {
+            scanType = Count;
+        } else if (scanTypeString == "snapshot") {
+            transactionOptions = mongocxx::options::transaction{};
+            auto readConcern = mongocxx::read_concern{};
+            readConcern.acknowledge_level(mongocxx::read_concern::level::k_majority);
+            transactionOptions->read_concern(readConcern);
+            scanType = Snapshot;
+        } else if (scanTypeString == "standard") {
+            scanType = Standard;
+        } else {
+            BOOST_THROW_EXCEPTION(InvalidConfigurationException("ScanType is invalid."));
+        }
+        if (generateCollectionNames) {
+            queryCollectionList = false;
+            BOOST_LOG_TRIVIAL(info) << " Generating collection names";
+            for (const auto& collectionName :
+                 distributeCollectionNames(collectionCount, threads, actor->_index)) {
+                collections.push_back(db[collectionName]);
+            }
+        } else {
+            queryCollectionList = true;
         }
     }
 };
 
+void collectionScan(genny::v1::ActorPhase<CollectionScanner::PhaseConfig>& config,
+                    std::vector<mongocxx::collection>& collections) {
+    /*
+     * Here we are either doing a snapshot collection scan
+     * or just a normal scan?
+     */
+    size_t docCount = 0;
+    size_t scanSize = 0;
+    bool scanFinished = false;
+    auto statTracker = config->scanOperation.start();
+    for (auto& collection : config->collections) {
+        auto filter = config->filterExpr ? config->filterExpr->evaluate()
+                                         : bsoncxx::document::view_or_value{};
+        auto docs = collection.find(filter);
+        /*
+         * Try-catch this as the collection may have been deleted.
+         * You can still do a find but it'll throw an exception when we iterate.
+         */
+        try {
+            for (auto& doc : docs) {
+                docCount += 1;
+                scanSize += doc.length();
+                if (config->documents != 0 && config->documents == docCount) {
+                    scanFinished = true;
+                    break;
+                }
+                if (config->scanSizeBytes != 0 && scanSize >= config->scanSizeBytes) {
+                    scanFinished = true;
+                    break;
+                }
+            }
+            if (scanFinished) {
+                break;
+            }
+        } catch (const mongocxx::operation_exception& e) {
+            auto exceptionsCaught = config->exceptionsCaught.start();
+            exceptionsCaught.addDocuments(1);
```

Comment:
I think start `exceptionsCaught` at the same time you start `statTracker` and just call `addDocuments(1)` inside this `catch` block. This way you know how long it took for the start of the iteration to occur and how many (in total) exceptions you got.

(Same thing for the other places where this is being done)